src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala

```diff
@@ -69,40 +85,213 @@ object FunctionRegistry {
         else {
           assert(priority != 0)
           Left(Ambiguous(name, typ, it))
+        }
       }
-    }
   }
 
   private def bind(name: String, typ: TypeTag, f: Fun) = {
-    lookup(name, typ) match {
-      case Right(existingBinding) =>
-        throw new RuntimeException(s"The name, $name, with type, $typ, is already bound as $existingBinding")
-      case _ =>
-        registry.updateValue(name, Seq.empty, (typ, f) +: _)
-    }
+    registry.updateValue(name, Seq.empty, (typ, f) +: _)
   }
 
-  def lookupField(ec: EvalContext)(typ: BaseType, name: String)(xAst: AST): Err[() => Any] =
-    lookup(name, MethodType(typ)).map {
-      case f: UnaryFun[_, _] => AST.evalCompose(ec, xAst)(f)
-      case f: OptionUnaryFun[_, _] => AST.evalFlatCompose(ec, xAst)(f)
-      case f =>
-        throw new RuntimeException(s"Internal hail error, bad binding in function registry for `$name' with argument type $typ: $f")
-    }
+  def lookupMethodReturnType(typ: Type, typs: Seq[Type], name: String): Err[Type] =
+    lookup(name, MethodType(typ +: typs: _*)).map(_.retType)
+
+  def lookupMethod(ec: EvalContext)(typ: Type, typs: Seq[Type], name: String)(lhs: AST, args: Seq[AST]): Err[() => Any] = {
+    require(typs.length == args.length)
+
+    val m = lookup(name, MethodType(typ +: typs: _*))
+    m.map {
+      case aggregator: Arity0Aggregator[_, _] =>
+        val localA = ec.a
+        val idx = localA.length
+        localA += null
+        ec.aggregations += ((idx, lhs.eval(ec), aggregator.ctor()))
+        () => localA(idx)
+
+      case aggregator: Arity1Aggregator[_, u, _] =>
+        val localA = ec.a
+        val idx = localA.length
+        localA += null
+
+        val u = args(0).eval(EvalContext())()
```

Comment:
Note, we're _calling_ the result of eval here at evaluation time.  This is a compile-time constant!  We could include the global context when it is available since it is known at compile-time, but (1) how do we get it here (do we include it in the EvalContext?) and (2) we use expressions in contexts where we don't have a global state (e.g. KeyTables).  Seems better this way to me.